juju/charm.py

```diff
@@ -82,19 +117,145 @@ def __init__(self, framework, key, metadata):
         super().__init__(framework, key)
         self.metadata = metadata
 
-        for role in ('requires', 'provides', 'peers'):
-            for relation_name in metadata.get(role, ()):
-                self.on.define_event(f'{relation_name}_relation_joined',
-                                     RelationJoinedEvent)
-                self.on.define_event(f'{relation_name}_relation_changed',
-                                     RelationChangedEvent)
-                self.on.define_event(f'{relation_name}_relation_departed',
-                                     RelationDepartedEvent)
-                self.on.define_event(f'{relation_name}_relation_broken',
-                                     RelationBrokenEvent)
-
-        for storage_name in metadata.get('storage', ()):
-            self.on.define_event(f'{storage_name}_storage_attached',
-                                 StorageAttachedEvent)
-            self.on.define_event(f'{storage_name}_storage_detaching',
-                                 StorageDetachingEvent)
+        for relation_name in self.metadata.relations:
+            self.on.define_event(f'{relation_name}_relation_joined', RelationJoinedEvent)
+            self.on.define_event(f'{relation_name}_relation_changed', RelationChangedEvent)
+            self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
+            self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
+
+        for storage_name in metadata.storage:
+            self.on.define_event(f'{storage_name}_storage_attached', StorageAttachedEvent)
+            self.on.define_event(f'{storage_name}_storage_detaching', StorageDetachingEvent)
+
+
+class CharmMetadata:
+    """Object containing the metadata for the charm.
+
+    Supported fields are:
+
+        * ``name``
+        * ``summary``
+        * ``description``
```

Comment:
That's somewhat longwinded without adding much value. I suggest cutting it down to something simpler and focusing on keeping only the bits that are worth calling out.

Last week we also got to an agreement to avoid adding implicit syntax to documentation until we feel like the value is worth it. The argument is that it's read many more times in the code itself than it is read elsewhere, so making it hard to read for people passing by the code is a bad trade-off. That's not such a big deal here, but it's more clear when we looking at things like ```:class:`RelationDefinition` ```, vs. simply `RelationDefinition`

To be clear, we can change that at some point, but I suggest at least for now focusing on making documentation excellent to read, and focusing less on the rich formatting of it.

Reply:
I'm used to writing doc strings for Python libraries using Sphinx to put docs up on RTD, and the `__init__` assigned fields wouldn't show up in that.  Similarly for the ``:class:`RelationDefintion` `` syntax, since that's the standard cross-ref link markup for Sphinx.  If that's not a concern for now, then you're right that this is only chaff that should be dropped.